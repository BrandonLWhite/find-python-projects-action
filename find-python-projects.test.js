const core = require("@actions/core");
const mapValues = require("lodash/mapValues.js");

const { run } = require("./find-python-projects.js");

describe("find-python-projects", () => {
  const infoMock = jest.spyOn(core, "info").mockImplementation();
  const getInputMock = jest.spyOn(core, "getInput").mockImplementation();
  const setFailedMock = jest.spyOn(core, "setFailed").mockImplementation();
  const setOutputMock = jest.spyOn(core, "setOutput").mockImplementation();

  const inputsDefaults = {
    "desired-export-paths": "",
  };
  let inputs = {};
  let outputs = {};

  beforeEach(() => {
    jest.clearAllMocks();
    inputs = {
      ...inputsDefaults,
    };
    outputs = {};

    // Mock the action's inputs
    getInputMock.mockImplementation((name) => {
      return inputs[name];
    });

    setOutputMock.mockImplementation((name, value) => {
      outputs[name] = value;
    });
  });

  it("finds projects", async () => {
    inputs["root-dir"] = "test-fixtures/multi-project";
    await run();
    expect(deserializeJsonValues(outputs)).toMatchSnapshot();
    expect(infoMock).toHaveBeenCalled();
  });

  it("Calls setFailed on error", async () => {
    inputs["root-dir"] = "test-fixtures/invalid-project";
    await run();
    expect(setFailedMock).toHaveBeenCalled();
  });

  it("Exports keys as instructed", async () => {
    inputs["desired-export-paths"] = "tool.export.me.please,what";
    inputs["root-dir"] = "test-fixtures/project-with-exports";
    await run();
    expect(deserializeJsonValues(outputs)).toMatchSnapshot();
    expect(infoMock).toHaveBeenCalled();
  });
});

function deserializeJsonValues(obj) {
  return mapValues(obj, (val) => JSON.parse(val));
}
