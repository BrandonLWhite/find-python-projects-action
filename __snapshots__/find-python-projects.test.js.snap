// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-python-projects finds projects 1`] = `
{
  "packageable-projects": [
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
        "package": "poetry bundle venv",
        "test": "poetry run pytest",
      },
      "directory": "test-fixtures/multi-project/project-2",
      "installCommand": "poetry install",
      "name": "sub-project-2",
      "packageCommand": "poetry bundle venv",
      "path": "test-fixtures/multi-project/project-2/pyproject.toml",
      "pythonVersion": "^3.9",
      "testCommand": "poetry run pytest",
    },
  ],
  "paths": [
    "test-fixtures/multi-project/project-1/pyproject.toml",
    "test-fixtures/multi-project/project-2/pyproject.toml",
    "test-fixtures/multi-project/project-4/pyproject.toml",
    "test-fixtures/multi-project/project-5/pyproject.toml",
    "test-fixtures/multi-project/project-6/pyproject.toml",
  ],
  "projects": [
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
      },
      "directory": "test-fixtures/multi-project/project-1",
      "installCommand": "poetry install",
      "name": "sub-project-1",
      "path": "test-fixtures/multi-project/project-1/pyproject.toml",
      "pythonVersion": "^3.9",
    },
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
        "package": "poetry bundle venv",
        "test": "poetry run pytest",
      },
      "directory": "test-fixtures/multi-project/project-2",
      "installCommand": "poetry install",
      "name": "sub-project-2",
      "packageCommand": "poetry bundle venv",
      "path": "test-fixtures/multi-project/project-2/pyproject.toml",
      "pythonVersion": "^3.9",
      "testCommand": "poetry run pytest",
    },
    {
      "commands": {
        "install": null,
      },
      "directory": "test-fixtures/multi-project/project-4",
      "installCommand": null,
      "name": "sub-project-4",
      "path": "test-fixtures/multi-project/project-4/pyproject.toml",
      "pythonVersion": "^3.9",
    },
    {
      "buildBackend": "pdm.backend",
      "commands": {
        "install": "pdm install",
        "test": "pdm run test",
      },
      "directory": "test-fixtures/multi-project/project-5",
      "installCommand": "pdm install",
      "name": "project-5",
      "path": "test-fixtures/multi-project/project-5/pyproject.toml",
      "pythonVersion": "==3.12.*",
      "testCommand": "pdm run test",
    },
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
        "test": "poe run test",
      },
      "directory": "test-fixtures/multi-project/project-6",
      "installCommand": "poetry install",
      "name": "sub-project-1",
      "path": "test-fixtures/multi-project/project-6/pyproject.toml",
      "pythonVersion": "^3.9",
      "testCommand": "poe run test",
    },
  ],
  "testable-projects": [
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
        "package": "poetry bundle venv",
        "test": "poetry run pytest",
      },
      "directory": "test-fixtures/multi-project/project-2",
      "installCommand": "poetry install",
      "name": "sub-project-2",
      "packageCommand": "poetry bundle venv",
      "path": "test-fixtures/multi-project/project-2/pyproject.toml",
      "pythonVersion": "^3.9",
      "testCommand": "poetry run pytest",
    },
    {
      "buildBackend": "pdm.backend",
      "commands": {
        "install": "pdm install",
        "test": "pdm run test",
      },
      "directory": "test-fixtures/multi-project/project-5",
      "installCommand": "pdm install",
      "name": "project-5",
      "path": "test-fixtures/multi-project/project-5/pyproject.toml",
      "pythonVersion": "==3.12.*",
      "testCommand": "pdm run test",
    },
    {
      "buildBackend": "poetry.core.masonry.api",
      "commands": {
        "install": "poetry install",
        "test": "poe run test",
      },
      "directory": "test-fixtures/multi-project/project-6",
      "installCommand": "poetry install",
      "name": "sub-project-1",
      "path": "test-fixtures/multi-project/project-6/pyproject.toml",
      "pythonVersion": "^3.9",
      "testCommand": "poe run test",
    },
  ],
}
`;
